src/doc_handle.rs

```diff
@@ -0,0 +1,46 @@
+use automerge as am;
+use futures::Stream;
+
+use crate::DocumentId;
+
+/// A handle to a document managed by the repository
+///
+/// This is effectively an async mutex around an automerge document
+#[derive(Clone)]
+pub struct DocHandle {}
+
+impl DocHandle {
+    pub fn id(&self) -> DocumentId {
+        todo!()
+    }
+
+    /// Wait for this document to be unlocked and then modify it
+    ///
+    /// Any changes made to the document by `F` will be enqueue to send to all
+    /// connected peers who are sharing this document
+    pub async fn modify<O, E, F: FnOnce(&am::Automerge) -> Result<O, E>>(
+        &mut self,
+        f: F,
+    ) -> Result<O, E> {
+        todo!()
+    }
+
+    /// Wait for a document to be unlocked and get a read only ref to it
+    pub async fn read(&self) -> am::Automerge {
+        todo!()
+    }
+
+    /// Wait for a change to occur to this document and return a read only reference to the
+    /// document if it did change
+    pub async fn changed(&self) -> &am::Automerge {
```

Comment:
I prefer the ["sync observer" concept](https://github.com/automerge/spanreed/blob/1727072171bf719687853e763f1e347a38160724/src/repo.rs#L244), because it is not tied to a doc handle. 

If the method is tied to a doc handle, I can imagine the user having to build their own observing task using a handle and calling `changed.await` in a loop. Then if they wanted to read or modify the doc concurrently, they'd have to use a clone of it. 

Finally, what I found useful about the sync observer, is that it could be a way to get a handle for a document that is synced over the network for which no handle had been previously acquired(for example if the repo starts syncing with a peer following an announcement of document).

Reply:
A sync observer API may be useful in addition. But I think being able to listen to an individual doc handle is crucial. Consider a websocket handler which is streaming updates to a specific document to a web frontend (such as in a collaborative text editor). In this case it would be inconvenient to only have a global sync observer, the user would need to manage their own state tracking all sync states and then somehow forwarding them to the websocket handler. That's a lot of boilerplate.

In general there are going to be parts of the application which are interested in changes to a specific document and forcing those application to ship around a bunch of shared state to communicate with the sync observer is something I think we should avoid as we are already doing that work anyway